openedx/core/djangoapps/safe_sessions/middleware.py

```diff
@@ -672,6 +691,10 @@ def _verify_user_unchanged(request, response, userid_in_session):
         if hasattr(request.user, 'real_user'):
             # If a view overrode the request.user with a masqueraded user, this will
             #   revert/clean-up that change during response processing.
+            #   Known places this is set:
+            #
+            #   - lms.djangoapps.courseware.masquerade::setup_masquerade
+            #   - openedx.core.djangoapps.content.learning_sequences.views::CourseOutlineView
```

Comment:
[non-blocking/future] I don't like that this code is so exceptional, and that these new comments are likely to become out-of-date. Could we simply use `mark_user_change_as_expected` instead? 
Possible remediation steps:
1. Add New Relic custom attribute to see when we are reaching this code. (Maybe we'll need to log a stacktrace later?)
2. Add calls to `mark_user_change_as_expected` until we no longer see transactions relying on this in New Relic.

Reply:
One thing that's nice about this is that (at the moment) it's more fine-grained -- it pops a stack and restores the user, then allows all the other checks to proceed. (But it's also strictly for session-type masquerading, which doesn't change the session itself.) Using the mark-change function currently skips over all the checks, which makes me a little sad.

Connecting this with the other discussion, it's possible that we could have the best of both worlds by starting to validate the expected new user ID (but still skip the remaining checks). Then ignoring `real_user` wouldn't have the downside of loss-of-precision any more.